### ISSUE TITLE
Tests are failing on Ubuntu 20.04 with GCC

### ISSUE THREAD

**Issue by kuhar** [GPUOpen-Drivers/llpc#1096]
Two shaderdb tests started to fail after switching from 18.04 to 20.04:

```
  LLPC_SHADERTEST :: shaderdb/PipelineCs_TestFetch2DMSFmaskBased_lit.pipe
  LLPC_SHADERTEST :: shaderdb/relocatable_shaders/PipelineVsFs_ShadowDescTableMissingFmask.pipe
```

Sample failure: https://github.com/GPUOpen-Drivers/llpc/runs/1537169613?check_suite_focus=true.

Sebastian suspects some memory corruption: https://github.com/GPUOpen-Drivers/llpc/issues/1092#issuecomment-743165624.

cc @s-perron @Flakebi @JaxLinAMD

**Comment by kuhar** [comment]
Local repro:

```
sudo docker build . --file docker/amdvlk.Dockerfile    \
               --build-arg BRANCH=dev                            \                                                        
               --build-arg CONFIG=Release                     \
               --build-arg ASSERTIONS=ON                    \
               --build-arg FEATURES="+gcc"                   \
               --build-arg GENERATOR=Ninja    
```

**Comment by s-perron** [comment]
I looked into this a little.  It seems like `PipelineState::findResourceNode` does not find the resource node for the shadow descriptor in the failing case.  Adding debugging output to that function changes the behaviour.  Nothing in that functions jumps out at me as incorrect.  I probably won't be looking into this unless help is needed.

**Comment by Flakebi** [comment]
I made some progress on this.
**tl;dr:** I suspect it is a bug in gcc’s tail recursive call optimization.

Single-stepping through the execution of `PipelineState::findResourceNode` showed that the failing runs with a gcc compiled amdllpc does never run the `if (nodeType == ResourceNodeType::DescriptorFmask …)` code. The assembly does not contain a check for `ResourceNodeType::DescriptorFmask`, which manifests itself as `cmp    $0x5,%…` in working binaries.

Adding a side-effect like `printf` into the function fix the code generation and make the test pass. In these cases, gcc still does not emit a call to `PipelineState::findResourceNode` but makes use of tail recursive call optimization and uses a jump to the function start (after prologue).

Changes that make the test pass:
- Compile with clang instead of gcc
- Compile with gcc from Ubuntu 18.04 (I didn’t try this but it worked on GitHub CI)
- Add a side-effect (`printf`) somewhere in the function before or inside `if (nodeType == ResourceNodeType::DescriptorFmask`
- Compile with `-fno-optimize-sibling-calls` (I recompiled only `PipelineState.cpp` with `-fno-optimize-sibling-calls`, the rest of the compiler stayed identical)

Changes that keep the test failing:
- Add a side-effect (`printf`) immediately before `return {nullptr, nullptr};`

For reference, I compiled in debug mode using gcc 9.3.0.

Any idea on how to progress from here? Should we add `-fno-optimize-sibling-calls` to our gcc compilation flags?

**Comment by kuhar** [comment]
@Flakebi thanks for looking into this.
I just tried upgrading gcc and g+++ to 10 and compiling under docker, but it fails in the same way. Very strange.

**Comment by kuhar** [comment]
I also tried updating libstdc++, glibc, and multilib, but that didn't help either.

**Comment by Flakebi** [comment]
Thanks for testing with gcc 10.

It’s gcc/compilation related only, so changing the standard library shouldn’t change anything. The emitted assembly for `PipelineState::findResourceNode` is missing the check for `ResourceNodeType::DescriptorFmask`.

The only two possibilities I see is either
1. through the tail recursion, gcc sees an optimization opportunity (maybe through ub), which ends up in an endless loop, allowing it to completely optimize out the `ResourceNodeType::DescriptorFmask` branch
2. or gcc has a bug that makes it see an optimization opportunity where there is none and it still ends up removing the branch.

**Comment by kuhar** [comment]
or 3. llpc is at fault and we get a non-local effect of some UB in the codebase

If conventional means don't help, I think using an external function that clobbers memory under preprocessor guards for gcc should be acceptable.

Reverting to older Ubuntu sounds like a much worse of an option to me and pretending like the issue doesn't exist, while it may affect real users.

**Comment by kuhar** [comment]
llvm's `benchmark.h` defines something like this:

```
// The DoNotOptimize(...) function can be used to prevent a value or
// expression from being optimized away by the compiler. This function is
// intended to add little to no overhead.
// See: https://youtu.be/nXaxk27zwlk?t=2441
#ifndef BENCHMARK_HAS_NO_INLINE_ASSEMBLY
template <class Tp>
inline BENCHMARK_ALWAYS_INLINE
void DoNotOptimize(Tp const& value) {
    asm volatile("" : : "r,m"(value) : "memory");
}

template <class Tp>
inline BENCHMARK_ALWAYS_INLINE void DoNotOptimize(Tp& value) {
#if defined(__clang__)
  asm volatile("" : "+r,m"(value) : : "memory");
#else
  asm volatile("" : "+m,r"(value) : : "memory");
#endif
}
```

**Comment by kuhar** [comment]
The latest nightly run passed with GCC.
https://github.com/GPUOpen-Drivers/llpc/actions/runs/425624017

Thanks for the fix, Sebastian.
